### ISSUE TITLE
Introduce black formatting in the codebase

### ISSUE THREAD

**Issue by gabrieldemarmiesse** [cython/cython#5127]
I did not see any other issue related to this.

Black has allowed many open sourced project I worked on to gain more contributors, notably because people don't fight over formatting, and they don't fight with the CI to match the style imposed by a tool that doesn't autocorrect the code (pycodestyle or flake8).

We have two ways to introduce black to this massive codebase if we wish to do so:
1) We ignore all existing files and impose black on new files to avoid backport issues and conflicts in PRs
2) We format the whole codebase at once and ignore the commit (for git blame) with https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html 

Are the maintainers open to such a move?

**Comment by da-woods** [comment]
I've definitely been doing it on new files (although I'm not immediately sure whether any PR with black formatting has yet been merged. Pattern matching will be mostly black-formatted when merged). It definitely isn't imposed yet.

Not sure about doing it on everything. I struggle to believe that the git blame ignore is completely effective. But I can see the appeal of "impose it and then ignore formatting issues forever".

I guess the question is "how many useful contributors are put off by formatting requirements?". And we probably can never know because they probably don't tell us but just don't come back.

**Comment by matusvalo** [comment]
> Not sure about doing it on everything. I struggle to believe that the git blame ignore is completely effective. But I can see the appeal of "impose it and then ignore formatting issues forever".

Other big projects succesfully migrated to black - e.g. django [1][2]:



The biggest benefit of using black is zero overhead for devs to manage code formatting. This is even better when pre-commit is introduced and new commit is added to PR by CI when black reformat the code - see pylint as an example [3]

Regarding the git blame it should be manageable [4]

[1] https://github.com/django/django/blob/1be7e36f85c927560e8c3c1eda05a7e43a66cd22/pyproject.toml#L5-L7
[2] https://groups.google.com/g/django-developers/c/wK2PzdGNOpQ
[3] https://results.pre-commit.ci/run/github/47671127/1668354497.A9rB447jQN23uvMfa9qLUQ
[4] https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html

**Comment by mattip** [comment]
As an observer from the side: the main argument for using black is "never argue about formatting again". But now the project, that rarely if ever needs to discuss formatting, has to deal with whether to use black or not. Personally, I would prefer Cython not to add a new tool with its own baggage of git work-arounds, version updates and CI complications.

**Comment by scoder** [comment]
I'm also reluctant to reformat the code base globally, although formatting really isn't always obvious, to no-one.

I think it's ok to use it for newly introduced files, I'm less convinced about the modules.

**Comment by gabrieldemarmiesse** [comment]
I'll make two pull requests, one with a huge commit + the ignore blame file, and one with the setup to impose black on new files. You'll be able to play with it, notably with the blame, and that should help push the discussion forward (and merge one of them if we agree)

**Comment by gabrieldemarmiesse** [comment]
I get errors like this 
```
Error compiling Cython file:
------------------------------------------------------------
...
          |name| is optional, and may be the name of the file being
          scanned or any other identifying string.
        """
        self.trace = 0

        self.buffer = ""
                      ^
------------------------------------------------------------

Cython/Plex/Scanners.py:92:22: str objects do not support coercion to unicode, use a unicode string literal instead (u'')
```
after black changes `self.buffer = u''` to `self.buffer = ""`.

I suspect this is because of the directive 
```
# cython: language_level=3str
```
does that mean that we can't apply back because the assumptions of black about strings are not valid in cython with this directive ?

**Comment by mattip** [comment]
Maybe this should all wait until after we drop python2.7. I think the potential for problems outweighs any perceived benefits from formatting this very complicated and old code base.

**Comment by eli-schwartz** [comment]
I avoid contributing to projects that mandate either black or pre-commit. The former because it results in unreadable code, the latter because it's hassle to set up locally and I object to remote CI applying non-`!fixup` commits that make history hard to follow.

**Comment by robertwb** [comment]
While I'm generally a fan of auto-formatters, care should be taken when applying them arbitrarily to a large, an existing codebase. (I also don't think it's that urgent because formatting arguments have not generally been a source of conflict in this project as they have been in some others.) And if we're going to open this can of worms, we should consider yapf as well.

**Comment by scoder** [comment]
> I avoid contributing to projects that mandate either black or pre-commit. The former because it results in unreadable code

I think that's exaggerating it. Yes, some code becomes less "but I wanted it so" when using an automatic code formatter. But from my experience, it's really rare that I'm so unhappy with the result that I consider it worth a manual change. Happens, ok. But even then, it's usually not too difficult to find a spelling on which me and the code formatter can agree. And that makes me rethink the way I've formatted the code, which is actually often a good thing.

**Comment by scoder** [comment]
IMHO, if we want to start using black (or maybe rather blue), Cython 3.1 would be the time to start. We'd be ripping out a lot of Py2 related code at that point, which already makes backports to 3.0 difficult, because they'd probably require adding a Py2 implementation or making the code itself Py2 compatible. Code formatting differences won't add much to that.

**Comment by eli-schwartz** [comment]
> I think that's exaggerating it.

The converting of consistent single quotes to consistent double quotes is meh and less "but I wanted it so".

The mangling of function signatures across multiple lines, and converting this:

``` python
def foo(....) -> ...:
    do_things()
```
to this:
```python
def foo(
    ...
) -> ...:
    do_things()
```

is ***literally*** unreadable to me, and it is indeed so bad that I'm so unhappy that I'd convert it all back.

Surely there must be a python code formatter that solves formatting woes without making code unreadable in the process. yapf exists, at least. I haven't personally tried it, so I cannot judge, unlike black which I've tried and judged. :P I haven't tried blue either (seems to be black, slightly modified for single quotes and some configuration?) Seems interesting.

Of course, my opinions as a rare contributor may not be all that representative. But there you have it, for what it's worth.

**Comment by merlinz01** [comment]
You should check out [ruff](https://astral.sh/ruff) for code formatting and linting.

**Comment by eli-schwartz** [comment]
Note that ruff is just another color of the bikeshed. Its main claim to fame is "Introducing, the one and only! A rewrite of black, but written in rust and a lot faster! Also we support your choice of single quotes, if you like, the same way several black forks such as `blue` do".

It was an intentional goal of ruff that there should be effectively no changes produced when reformatting black code using ruff. There are a small handful of minor differences. Nobody who wants a formatter will be overly concerned about whether to use black or ruff. Anyone who dislikes black is *exceedingly* unlikely to be appeased by ruff, though.

**Comment by scoder** [comment]
I don't think we will do this for 3.1.
